### PR TITLE
Rework throttled refresh

### PR DESCRIPTION
**Pull request by BurningTreeC**
This PR reworks the throttled refresh so that it works for the styleRefresh, too.
It adds two switches so that we can control if we want only the styleRefresh to throttle or only the mainRefresh to throttle.
It also makes the styleRefresh in new Windows equal to the styleRefresh in the main window.
Also in new windows the throttling is added, the same way as in the main window.

### CODE REVIEW

File: core/modules/startup/render.js [hunk 1]
```diff
@@ -51,9 +51,11 @@ exports.startup = function() {
...
 	document.head.insertBefore($tw.styleElement,document.head.firstChild);
 	// Prepare refresh mechanism
-	var deferredChanges = Object.create(null),
-		timerId,
-		throttledRefreshFn = function(changes,callback,mainCondition,styleCondition) {
+	var mainDeferredChanges = Object.create(null),
+		styleDeferredChanges = Object.create(null),
+		mainTimerId,
+		styleTimerId,
+		throttledRefreshFn = function(changes,deferredChanges,timerId,throttledRefresh,callback,mainCondition,styleCondition) {
```
Comment (pmario) [inline core/modules/startup/render.js:58]: I think it would be better to add the new function parameters as an options-object. So the function can be extended without changing the function signature again. The more parameters a function has the more complicated it is to call it. 

So imo you should go with

```
throttledRefreshFn = function(changes,callback,mainCondition,styleCondition,options) {
options = options || {}
...
```

Comment (BurningTreeC) [reply core/modules/startup/render.js:58]: Thanks @pmario ! Will do!

Comment (BurningTreeC) [reply core/modules/startup/render.js]: @pmario - I've moved almost everything into the options object

**Comment by BurningTreeC** [comment]
This now works and deferres the changes for both styleRefresh and mainRefresh in both the main window and new windows

**Comment by BurningTreeC** [comment]
@Jermolene @pmario @saqimtiaz here's a build with performance instrumentation enabled:

https://rework-throttled-refresh.tiddlyhost.com/

**Comment by pmario** [comment]
>... here's a build with performance instrumentation enabled:

Seems to be slightly faster than the tiddlywiki/prerelease

**Comment by Leilei332** [comment]
Actually I wonder why opening `$:/TagManager` and `$:/Manager` is especially slow when there are many tiddlers or tags.

**Comment by BurningTreeC** [comment]
> Actually I wonder why opening `$:/TagManager` and `$:/Manager` is especially slow when there are many tiddlers or tags.

Hello @Leilei332 - do you notice this in general or in this PR?

**Comment by Leilei332** [comment]
> > Actually I wonder why opening `$:/TagManager` and `$:/Manager` is especially slow when there are many tiddlers or tags.
> 
> Hello @Leilei332 - do you notice this in general or in this PR?

This is a general problem in tiddlywiki. In this PR it is a little quicker.

Tag Manager (PR):

```
performance: styleRefresh: 8.00ms
performance: mainRefresh: 1103.00ms
```

Tag Manager (prerelease):

```
performance: styleRefresh: 13.00ms
performance: mainRefresh: 1122.00ms
```

Manager (PR):

```
performance: styleRefresh: 17.00ms
performance: mainRefresh: 513.00ms
```

Manager (prerelease):

```
performance: styleRefresh: 16.00ms
performance: mainRefresh: 514.00ms
```

**Comment by Leilei332** [comment]
I think the two tiddlers should be improved because its performance is slow on mobile phones. The result is the same (performance in this PR is a little faster than prerelease)

This PR:



Prerelease:

**Comment by Jermolene** [comment]
Thanks @BurningTreeC this is promising, but I don't think we're ready to include it in v5.3.6.

One issue is that the code of render.js is now pretty gnarly and hard to follow. I think we need to consider refactoring the rendering and refreshing into a generic class that can be used both for styles and body content. In a separate branch I've been exploring the idea with a new "wikifier" class. See [this commit](https://github.com/TiddlyWiki/TiddlyWiki5/commit/467a1a47cc01447bfa6068993f047c4e6de7bb8d#diff-5cf46a93ddbb41a3c5f8ac1d7ea3144d322fe5027ae2368da32c4f1e154d3095). It will need to be extended to allow it to replace all the occurrences of rendering/refreshing in the system.

**Comment by BurningTreeC** [comment]
> I think we need to consider refactoring the rendering and refreshing into a generic class that can be used both for styles and body content.

Hi @Jermolene , I believe I did just that

File: core/modules/utils/refresh-utils.js [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+title: $:/core/modules/utils/refresh-utils.js
+type: application/javascript
+module-type: utils
+
+Shared utilities for managing refresh operations with throttling support
+
+\*/
+
+/*jslint node: true, browser: true */
+/*global $tw: false */
```
Comment (Leilei332) [inline core/modules/utils/refresh-utils.js:11]: These comments are also not needed now.
